test: check that modules backed by read-only files can load (win32)

billti opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

This is the test-case for the issue at https://github.com/nodejs/node/issues/20112 . This fails with the issue present, and passes with the fix applied. 

Note: Even before any changes, `vcbuild test` gave a ton of failures for me, so not sure if that is my machine or something more fundamental. This test behaves as expected though.

CC @vsemozhetbyt, @cjihrig, @richardlau

Trott commented:
Hi, @billti! Thanks for this!

@nodejs/testing PTAL

Trott commented:
I guess @nodejs/platform-windows too...

Trott commented on test/parallel/test-module-readonly.js:
```diff
@@ -0,0 +1,35 @@
+
+if (process.platform == 'win32') {
+    // Create readOnlyMod.js and set to read only
+    const readOnlyMod = fixtures.path('readOnlyMod');
```
I think the sum total of the change would be:

* require `tmpdir` instead of `fixtures`
* call `tmpdir.refresh()` once
* Use `tmpdir.path()` instead of `fixtures.path()`

billti commented on test/parallel/test-module-readonly.js:
```diff
@@ -0,0 +1,35 @@
+    cp.execSync(`icacls.exe "${readOnlyModFullPath}" /remove "%USERNAME%" /inheritance:e`);
+
+    // Delete the file
+    fs.unlinkSync(readOnlyModFullPath);
```
I just pushed a change to use tmpdir. I figured cleaning up is safer than having another run try and overwrite the same file (permissions shouldn't conflict as it should be a different tmpdir if under a different account, but just to be safe...).

Trott commented:
Whoops! Not `os.tmpdir()`. Use `common.tmpdir()`! https://github.com/nodejs/node/tree/master/test/common#tmpdir-module

Or look at another test to see some example code: https://github.com/nodejs/node/blob/bf6ce47259ebfe9f6349c223cc1735726aee062d/test/parallel/test-fs-write-stream-close-without-callback.js#L7-L10

billti commented:
I switched to the tmpdir requested.

I added the test for the v10.x branch as from the linked issue I thought that was what is being targeted ASAP in order to get the 10.0 release done. I can target whatever branch you want though. Just let me know.

vsemozhetbyt commented:
Thank you!
There are som linting nits in the test, You can quickly lint your file in this way from the project root folder:
```sh
node tools/node_modules/eslint/bin/eslint.js test/parallel/test-module-readonly.js
```
Currently, you will get:
<details>
<summary>Some errors:</summary>

```sh
j:\temp\_git\node-fork\test\parallel\1test-module-readonly.js
   1:1   error  Mandatory module "common" must be loaded       node-core/required-modules
   1:1   error  Use the global form of 'use strict'            strict
   8:22  error  Expected '===' and instead saw '=='            eqeqeq
   9:1   error  Expected indentation of 2 spaces but found 4   indent
  10:1   error  Expected indentation of 2 spaces but found 4   indent
  11:1   error  Expected indentation of 2 spaces but found 4   indent
  12:1   error  Expected indentation of 2 spaces but found 4   indent
  12:47  error  Strings must use singlequote                   quotes
  13:1   error  Expected indentation of 2 spaces but found 4   indent
  13:43  error  Strings must use singlequote                   quotes
  15:1   error  Expected indentation of 2 spaces but found 4   indent
  15:1   error  Line 15 exceeds the maximum line length of 80  max-len
  16:1   error  Expected indentation of 2 spaces but found 4   indent
  16:1   error  Line 16 exceeds the maximum line length of 80  max-len
  17:1   error  Expected indentation of 2 spaces but found 4   indent
  18:1   error  Expected indentation of 2 spaces but found 4   indent
  20:1   error  Expected indentation of 2 spaces but found 4   indent
  21:1   error  Expected indentation of 2 spaces but found 4   indent
  22:1   error  Expected indentation of 4 spaces but found 8   indent
  23:1   error  Expected indentation of 4 spaces but found 8   indent
  23:15  error  'mymod' is assigned a value but never used     no-unused-vars
  24:1   error  Expected indentation of 2 spaces but found 4   indent
  24:7   error  Expected space(s) after "catch"                keyword-spacing
  25:1   error  Expected indentation of 4 spaces but found 8   indent
  26:1   error  Expected indentation of 2 spaces but found 4   indent
  28:1   error  Expected indentation of 2 spaces but found 4   indent
  29:1   error  Expected indentation of 2 spaces but found 4   indent
  29:1   error  Line 29 exceeds the maximum line length of 80  max-len
  31:1   error  Expected indentation of 2 spaces but found 4   indent
  32:1   error  Expected indentation of 2 spaces but found 4   indent
  34:1   error  Expected indentation of 2 spaces but found 4   indent
  34:17  error  Use assert.ifError(except) instead             node-core/prefer-assert-iferror

✖ 32 problems (32 errors, 0 warnings)
  24 errors, 0 warnings potentially fixable with the `--fix` option.
```
</details><br>
About `Mandatory module "common" must be loaded` you can read in https://github.com/nodejs/node/blob/master/doc/guides/writing-tests.md

bnoordhuis commented on test/parallel/test-module-readonly.js:
```diff
@@ -0,0 +1,36 @@
+const tmpdir = require('../common/tmpdir');
+tmpdir.refresh();
+
+if (process.platform == 'win32') {
```
`common.isWindows`

cjihrig replied:
Since this is currently a Windows only test, you can do:

```
if (!common.isWindows) {
  common.skip('test only runs on Windows');
}
```

This will skip the test on other platforms and exit the process.

bnoordhuis commented on test/parallel/test-module-readonly.js:
```diff
@@ -0,0 +1,36 @@
+    fs.writeFileSync(readOnlyModFullPath, "module.exports = 42;");
+
+    // Removed any inherited ACEs, and any explicitly granted ACEs for the current user
+    cp.execSync(`icacls.exe "${readOnlyModFullPath}" /inheritance:r /remove "%USERNAME%"`);
```
Long line here, whitespace/indent errors elsewhere.  Can you run `make lint`?

richardlau replied:
As this is a Windows test the equivalent is `vcbuild lint` (or `vcbuild lint-js` to just lint the JavaScript files).

richardlau commented:
Adding the blocked label as this test is expected to fail (even when the lint issues have been addressed) until we get a libuv with https://github.com/libuv/libuv/pull/1800.

cjihrig commented on test/parallel/test-module-readonly.js:
```diff
@@ -0,0 +1,36 @@
+    // Delete the file
+    fs.unlinkSync(readOnlyModFullPath);
+
+    if (except) throw except;
```
Instead of an `if`, we'd ideally want an assertion, right?

richardlau replied:
e.g., `assert.ifError(except)` https://nodejs.org/dist/latest-v9.x/docs/api/assert.html#assert_assert_iferror_value

Trott commented:
`vcbuild lint` will flag a bunch of stuff on this file (or at least should) like needing a `'use strict';` and requiring the test `common` module. But I think those will be easy fixes.

Trott commented:
(Sorry for all the style nits that are getting picked. We try to put as many of them into the linter as possible rather than having them come up in PR review. Thanks for your patience with it all!)

billti commented:
No problem. I think I have all the issues addressed. Just trying to run all the tests/linting now. I'll open a new pull request against 'master' when it's ready, and link & close this one when done.

billti commented:
Opened new pull request against master in https://github.com/nodejs/node/pull/20138 which should address all comments (so far :wink:).
